### PR TITLE
feat(agents): granite agent

### PR DESCRIPTION
**Pull request by michael-desmond**
### Which issue(s) does this pull-request address?

partially fixes: #145

### Description

GraniteBeeAgent and runner that implements the granite tool pattern. 

### Checklist

- [x] I have read the [contributor guide](https://github.com/i-am-bee/bee-agent-framework/blob/main/CONTRIBUTING.md)
- [x] Linting passes: `yarn lint` or `yarn lint:fix`
- [x] Formatting is applied: `yarn format` or `yarn format:fix`
- [x] Unit tests pass: `yarn test:unit`
- [x] E2E tests pass: `yarn test:e2e`
- [ ] Tests are included 
- [x] Documentation is changed or added
- [x] Commit messages and PR title follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)

### CODE REVIEW

**Review by Tomas2D — COMMENTED** [review]
I just made some changes to the `main` branch and rebased + simplified your branch. 

Now we can discuss how to even more simplify the `_run` method.

**Review by Tomas2D — COMMENTED** [review]
I even simplified the implementation. Please check and update if needed.

**Comment by michael-desmond** [comment]
This looks good to me, I did some testing and the prompt being sent to the LLM looks correct. For completeness I added a granite bee example in the examples folder.

File: examples/agents/bee_granite.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import { parseEnv } from "bee-agent-framework/internals/env";
+
+// Ensure that you have added your WatsonX credentials in .env
+const llm = WatsonXChatLLM.fromPreset("ibm/granite-3-8b-instruct", {
+  apiKey: parseEnv("WATSONX_API_KEY", z.string()),
+  projectId: parseEnv("WATSONX_PROJECT_ID", z.string()),
+  parameters: {
+    decoding_method: "greedy",
+    max_new_tokens: 1024,
+  },
```
Comment (Tomas2D) [inline examples/agents/bee_granite.ts:19]: You are missing a stop sequence.

File: src/agents/granite/runner.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                description: tool.description.replaceAll("\n", ".").replace(/\.$/, "").concat("."),
+                parameters: R.omit(await tool.getInputJsonSchema(), [
+                  "minLength",
+                  "maxLength",
+                  "$schema",
+                ]),
+              })),
+            ),
+            null,
+            4,
```
Comment (Tomas2D) [inline src/agents/granite/runner.ts:49]: I would consider using `2` instead of `4`.

**Review by Tomas2D — COMMENTED** [review]
Could you please also update the documentation?

File: examples/agents/bee_granite.ts (hunk 1 above)
Comment (michael-desmond) [reply examples/agents/bee_granite.ts]: I moved the max_new_tokens into granite8b preset for watsonx [src/adapters/watsonx/chatPreset.ts](https://github.com/i-am-bee/bee-agent-framework/pull/159/files#diff-aadf764b9e5cf6016898f56145c827099fa6cc300339667531a2eb4cc6daee65) so I no longer need to configure the parameters in the example.

File: src/agents/granite/runner.ts (hunk 2 above)
Comment (michael-desmond) [reply src/agents/granite/runner.ts]: Done!

**Review by Tomas2D — APPROVED** [review]
Just a few notes. Then we can merge 🚀

File: src/agents/granite/agent.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ */
+
+import { BeeRunInput, BeeRunOptions, BeeRunOutput } from "@/agents/bee/types.js";
+import { GetRunContext } from "@/context.js";
+import { BeeAgent, BeeInput } from "@/agents/bee/agent.js";
+import { GraniteAgentRunner } from "./runner.js";
+import { GraniteBeeAssistantPrompt, GraniteBeeSystemPrompt } from "@/agents/granite/prompts.js";
+import { BaseMessage } from "@/llms/primitives/message.js";
+
+const TOOL_RESPONSE_ROLE = "tool_response";
```
[the remark this answers is not in the document]
Comment (michael-desmond) [reply src/agents/granite/agent.ts]: Done

File: src/agents/granite/runner.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import { BaseMessage, Role } from "@/llms/primitives/message.js";
+import { BeeRunOptions } from "@/agents/bee/types.js";
+import { AnyTool } from "@/tools/base.js";
+import { BeeInput } from "@/agents/bee/agent.js";
+import * as R from "remeda";
+import { BeeAgentRunner } from "@/agents/bee/runner.js";
+import { BeeParserInput } from "@/agents/bee/parser.js";
+import { isEmpty } from "remeda";
+
+const AVAILABLE_TOOLS_ROLE = "available_tools";
```
[the remark this answers is not in the document]
Comment (michael-desmond) [reply src/agents/granite/runner.ts]: Done

File: src/adapters/watsonx/chatPreset.ts [hunk 5]
```diff
@@ -30,6 +30,7 @@ export const WatsonXChatLLMPreset = {
       base: {
         parameters: {
           decoding_method: "greedy",
+          max_new_tokens: 512,
```
[the remark this answers is not in the document]
Comment (michael-desmond) [reply src/adapters/watsonx/chatPreset.ts:33]: I looked into the other providers. 
- For Bam granite is not included in the presets
- IBMvLLM defaults to unlimited max tokens

**Comment by michael-desmond** [comment]
I addressed the final few changes and added some docs re. granite 3.0 support. I kept the docs minimal.

**Review by Tomas2D — APPROVED** [review]
Fix lint and we can merge.

**Pull request merged**